mesonpy/_editable.py

```diff
@@ -2,183 +2,298 @@
+    def __init__(self, name: str, path: str, tree: Node):
+        super().__init__(name, path)
+        self._tree = tree
+
+    def get_resource_reader(self, name: str) -> TraversableResources:
+        return MesonpyReader(name, self._tree)
+
+
+LOADERS = [
+    (ExtensionFileLoader, tuple(importlib.machinery.EXTENSION_SUFFIXES)),
+    (SourceFileLoader, tuple(importlib.machinery.SOURCE_SUFFIXES)),
+    (SourcelessFileLoader, tuple(importlib.machinery.BYTECODE_SUFFIXES)),
+]
+
+
+def build_module_spec(cls: type, name: str, path: str, tree: Optional[Node]) -> importlib.machinery.ModuleSpec:
+    loader = cls(name, path, tree)
+    spec = importlib.machinery.ModuleSpec(name, loader, origin=path)
+    spec.has_location = True
+    if loader.is_package(name):
+        spec.submodule_search_locations = []
+    return spec
+
+
+class Node(NodeBase):
+    def __missing__(self, key: str) -> Node:
+        value = self[key] = Node()
+        return value
+
+    def __setitem__(self, key: Union[str, Tuple[str, ...]], value: Union[Node, str]) -> None:
+        n = self
+        if isinstance(key, tuple):
+            for k in key[:-1]:
+                n = n[k]  # type: ignore
+            key = key[-1]
+        dict.__setitem__(n, key, value)
+
+    def __getitem__(self, key: Union[str, Tuple[str, ...]]) -> Union[Node, str]:
+        n = self
+        if isinstance(key, tuple):
+            for k in key[:-1]:
+                n = n[k]  # type: ignore
+            key = key[-1]
+        return dict.__getitem__(n, key)
+
+    def get(self, key: Union[str, Tuple[str, ...]]) -> Optional[Union[Node, str]]:  # type: ignore[override]
+        n = self
+        if isinstance(key, tuple):
+            for k in key[:-1]:
+                n = dict.get(n, k)  # type: ignore
+                if n is None:
+                    return None
+            key = key[-1]
+        return dict.get(n, key)
+
+
+def walk(root: str, path: str = '') -> Iterator[pathlib.Path]:
+    with os.scandir(os.path.join(root, path)) as entries:
+        for entry in entries:
+            if entry.is_dir():
+                yield from walk(root, os.path.join(path, entry.name))
+            else:
+                yield pathlib.Path(path, entry.name)
+
+
+def collect(install_plan: Dict[str, Dict[str, Any]]) -> Node:
+    tree = Node()
+    for key, data in install_plan.items():
+        for src, target in data.items():
+            path = pathlib.Path(target['destination'])
+            if path.parts[0] in {'{py_platlib}', '{py_purelib}'}:
+                if key == 'install_subdirs' and os.path.isdir(src):
+                    for entry in walk(src):
+                        tree[(*path.parts[1:], *entry.parts)] = os.path.join(src, *entry.parts)
+                else:
+                    tree[path.parts[1:]] = src
+    return tree
+
+
+class MesonpyMetaFinder(importlib.abc.MetaPathFinder):
+    def __init__(self, names: Set[str], path: str, args: List[str], verbose: bool):
```

Comment:
Keep the default argument for `verbose`. There's no point in us having to manually pass it in tests.